applications/IgaApplication/custom_conditions/base_discrete_condition.h

```diff
@@ -0,0 +1,319 @@
+        const Variable<Vector>& rVariable,
+        Vector& rOutput,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+
+    /********************************************************************/
+    /*    SetValuesOnIntegrationPoints                                   */
+    /********************************************************************/
+    /**
+    * @brief Set a Constitutive Law Value on the Element
+    * @param rVariable The variable we want to set
+    * @param rValues The values to set in the integration points
+    * @param rCurrentProcessInfo the current process info instance
+    */
+    void SetValueOnIntegrationPoints(
+        const Variable<Vector>& rVariable,
+        std::vector<Vector>& rValues,
+        const ProcessInfo& rCurrentProcessInfo
+    ) override;
+
+
+
+    /// Turn back information as a string.
+    std::string Info() const override
+    {
+        std::stringstream buffer;
+        buffer << "\"BaseDiscreteCondition\" #" << Id();
+        return buffer.str();
+    }
+
+    /// Print information about this object.
+    void PrintInfo(std::ostream& rOStream) const override
+    {
+        rOStream << "\"BaseDiscreteCondition\" #" << Id();
+    }
+
+    /// Print object's data.
+    void PrintData(std::ostream& rOStream) const {
+        pGetGeometry()->PrintData(rOStream);
+    }
+    ///@}
+
+protected:
+
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    /**
+    * This functions calculates both the RHS and the LHS
+    * @param rLeftHandSideMatrix: The LHS
+    * @param rRightHandSideVector: The RHS
+    * @param rCurrentProcessInfo: The current process info instance
+    * @param CalculateStiffnessMatrixFlag: The flag to set if compute the LHS
+    * @param CalculateResidualVectorFlag: The flag to set if compute the RHS
+    */
+    virtual void CalculateAll(
+        MatrixType& rLeftHandSideMatrix,
+        VectorType& rRightHandSideVector,
+        ProcessInfo& rCurrentProcessInfo,
+        const bool CalculateStiffnessMatrixFlag,
+        const bool CalculateResidualVectorFlag);
+
+    /**
+    * CalculateJacobian computes the mapping from Geometry to Parameter Space.
+    *
+    * @param[in] DN_De derivatives of shape functions in two directions.
+    * @param[out] Jacobian calculated Jacobian. 
+    * @param[in] rWorkingSpaceDimension GeometrySpace coordinates. For surfaces default 3.
+    * @param[in] rLocalSpaceDimension ParameterSpace coordinates For surfaces default 2.
+    */
+    void CalculateJacobian(const Matrix& DN_De,
+        Matrix& Jacobian,
+        const int rWorkingSpaceDimension = 3,
+        const int rLocalSpaceDimension = 2);
+
+    /** Gets the number of dofs per node.
+    * If a derived condition has more degrees of freedom this function has to be overridden
+    *
+    * @return Number of dofs per node.
+    */
+    virtual inline std::size_t DofsPerNode() const
+    {
+        return 3;
```

Comment:
In the `IgaBaseElement` I used a template for this because it was the only way in which the compiler completely inlined all the code without function calls (it will be called a lot of times). But I also never managed to measure if there is a difference for real-life-problems...